File: jlpt/native_android/app/src/main/java/com/jlpt/android/data/remote/api/ExamApi.kt
package com.jlpt.android.data.remote.api

import com.jlpt.android.data.remote.dto.AttemptResponseDto
import com.jlpt.android.data.remote.dto.AttemptResultResponseDto
import com.jlpt.android.data.remote.dto.ExamDetailResponseDto
import com.jlpt.android.data.remote.dto.ExamSummaryResponseDto
import com.jlpt.android.data.remote.dto.SubmitAnswerRequestDto
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

interface ExamApi {
    @GET("api/exams")
    suspend fun listExams(
        @Query("level") level: String,
        @Query("skill") skill: String? = null
    ): List<ExamSummaryResponseDto>

    @GET("api/exams/{id}")
    suspend fun getExam(@Path("id") id: Long): ExamDetailResponseDto

    @POST("api/exams/{id}/attempts")
    suspend fun startAttempt(@Path("id") id: Long): AttemptResponseDto

    @GET("api/attempts/{id}")
    suspend fun getAttempt(@Path("id") id: Long): AttemptResponseDto

    @PUT("api/attempts/{id}/answers")
    suspend fun submitAnswer(@Path("id") id: Long, @Body request: SubmitAnswerRequestDto)

    @POST("api/attempts/{id}/submit")
    suspend fun submitAttempt(@Path("id") id: Long): AttemptResultResponseDto

    @GET("api/attempts/{id}/result")
    suspend fun getAttemptResult(@Path("id") id: Long): AttemptResultResponseDto

    @GET("api/users/me/attempts")
    suspend fun getMyAttempts(): List<AttemptResponseDto>
}

File: jlpt/native_ios/Sources/Data/API/ExamAPI.swift
import Foundation

final class ExamAPI {
    private let client: APIClient
    init(client: APIClient) { self.client = client }

    func listExams(level: String, skill: String?) async throws -> [ExamSummaryResponseDTO] {
        var query = [URLQueryItem(name: "level", value: level)]
        if let skill { query.append(URLQueryItem(name: "skill", value: skill)) }
        return try await client.send(Endpoint(path: "api/exams", query: query))
    }

    func getExam(id: Int64) async throws -> ExamDetailResponseDTO {
        try await client.send(Endpoint(path: "api/exams/\(id)"))
    }

    func startAttempt(examId: Int64) async throws -> AttemptResponseDTO {
        try await client.send(Endpoint(path: "api/exams/\(examId)/attempts", method: .post))
    }

    func getAttempt(id: Int64) async throws -> AttemptResponseDTO {
        try await client.send(Endpoint(path: "api/attempts/\(id)"))
    }

    func submitAnswer(attemptId: Int64, request: SubmitAnswerRequestDTO) async throws {
        try await client.sendNoContent(Endpoint(
            path: "api/attempts/\(attemptId)/answers",
            method: .put,
            body: Endpoint.withJSONBody(request, encoder: client.encoder)
        ))
    }

    func submitAttempt(attemptId: Int64) async throws -> AttemptResultResponseDTO {
        try await client.send(Endpoint(path: "api/attempts/\(attemptId)/submit", method: .post))
    }

    func getAttemptResult(attemptId: Int64) async throws -> AttemptResultResponseDTO {
        try await client.send(Endpoint(path: "api/attempts/\(attemptId)/result"))
    }

    func getMyAttempts() async throws -> [AttemptResponseDTO] {
        try await client.send(Endpoint(path: "api/users/me/attempts"))
    }
}
